lib/_http_client.js

```diff
@@ -62,9 +62,16 @@ function ClientRequest(options, cb) {
   var defaultAgent = options._defaultAgent || Agent.globalAgent;
   if (agent === false) {
     agent = new defaultAgent.constructor();
-  } else if ((agent === null || agent === undefined) &&
-             typeof options.createConnection !== 'function') {
-    agent = defaultAgent;
+  } else if (agent === null || agent === undefined) {
+    if (typeof options.createConnection !== 'function') {
+      agent = defaultAgent;
+    }
+    // Explicitly pass through this statement as agent will not be used
+    // when createConnection is provided.
+  } else if (!(agent instanceof Agent.Agent)) {
```

Comment:
How about userland `Agent`? Those Agent Class maybe not inherits from `Agent.Agent` and now will be all fails  to use them. e.g.: TunnelingAgent  https://github.com/request/tunnel-agent/blob/master/index.js#L47